receive_stream.go

```diff
@@ -272,13 +272,22 @@ func (s *receiveStream) onClose(offset protocol.ByteCount) {
 
 func (s *receiveStream) SetReadDeadline(t time.Time) error {
 	s.mutex.Lock()
-	oldDeadline := s.readDeadline
-	s.readDeadline = t
+	s.deadline = t
 	s.mutex.Unlock()
-	// if the new deadline is before the currently set deadline, wake up Read()
-	if t.Before(oldDeadline) {
+	if s.deadline.IsZero() { // skip if there's no deadline to set
 		s.signalRead()
+		return nil
+	}
+	// Lazily initialize the deadline timer.
+	if s.deadlineTimer == nil {
```

Comment:
Does this not need to be protected by the mutex? (same in the other commit)

Reply:
Good catch. I think we can just use a `defer` here, since the deadline calls won't be used that often.

Reply:
If you want it to be thread safe, then `sync.Once` might be the best option.

Reply:
It would, if we didn't have the `deadline`, which also needs to be thread-safe.